crates/uopool/src/validate/mod.rs

```diff
@@ -0,0 +1,110 @@
+    sanity::SanityCheckError,
+    simulation::{CodeHash, SimulationCheckError, NUMBER_LEVELS},
+    uopool::ValidationError,
+    Chain, UserOperation, UserOperationHash,
+};
+use enumset::{EnumSet, EnumSetType};
+use ethers::{providers::Middleware, types::U256};
+use std::sync::Arc;
+
+pub mod sanity;
+pub mod simulation;
+pub mod simulation_trace;
+mod utils;
+pub mod validator;
+
+#[derive(Debug, Clone, Default)]
+pub struct UserOperationValidationOutcome {
+    pub prev_hash: Option<UserOperationHash>,
+    pub pre_fund: U256,
+    pub verification_gas_limit: U256,
+    // Simulation
+    pub valid_after: Option<U256>,
+    // Simulation trace
+    pub code_hashes: Option<Vec<CodeHash>>,
+}
+
+#[derive(EnumSetType, Debug)]
+pub enum UserOperationValidatorMode {
+    Sanity,
+    Simulation,
+    SimulationTrace,
+}
+
+#[async_trait::async_trait]
+pub trait UserOperationValidator: Send + Sync {
+    async fn validate_user_operation(
+        &self,
+        uo: &UserOperation,
+        mempool: &MempoolBox<VecUo, VecCh>,
+        reputation: &ReputationBox<Vec<ReputationEntry>>,
+        mode: EnumSet<UserOperationValidatorMode>,
+    ) -> Result<UserOperationValidationOutcome, ValidationError>;
+}
+
+pub struct SanityHelper<'a, M: Middleware + 'static> {
+    mempool: &'a MempoolBox<VecUo, VecCh>,
+    reputation: &'a ReputationBox<Vec<ReputationEntry>>,
+    eth_client: Arc<M>,
+    entry_point: EntryPoint<M>,
+    chain: Chain,
+}
+
+#[async_trait::async_trait]
+pub trait SanityCheck<M: Middleware>: Send + Sync {
+    async fn check_user_operation(
+        &self,
+        uo: &UserOperation,
+        helper: &mut SanityHelper<M>,
+    ) -> Result<(), SanityCheckError>;
+}
+
+pub struct SimulationHelper<'a, M: Middleware + 'static> {
+    mempool: &'a MempoolBox<VecUo, VecCh>,
+    reputation: &'a ReputationBox<Vec<ReputationEntry>>,
+    eth_client: Arc<M>,
+    entry_point: EntryPoint<M>,
+    chain: Chain,
+    simulate_validation_result: &'a SimulateValidationResult,
+    valid_after: Option<U256>,
+}
+
+#[async_trait::async_trait]
+pub trait SimulationCheck<M: Middleware>: Send + Sync {
+    async fn check_user_operation(
+        &self,
+        uo: &UserOperation,
+        helper: &mut SimulationHelper<M>,
+    ) -> Result<(), SimulationCheckError>;
+}
+
+pub struct SimulationTraceHelper<'a, M: Middleware + 'static> {
+    mempool: &'a MempoolBox<VecUo, VecCh>,
+    reputation: &'a ReputationBox<Vec<ReputationEntry>>,
+    eth_client: Arc<M>,
+    entry_point: EntryPoint<M>,
+    chain: Chain,
+    simulate_validation_result: &'a SimulateValidationResult,
+    js_trace: &'a JsTracerFrame,
+    stake_info: Option<[StakeInfo; NUMBER_LEVELS]>,
+    code_hashes: Option<Vec<CodeHash>>,
+}
+
+#[async_trait::async_trait]
+pub trait SimulationTraceCheck<M: Middleware>: Send + Sync {
+    async fn check_user_operation(
+        &self,
+        uo: &UserOperation,
+        helper: &mut SimulationTraceHelper<M>,
```

Comment:
The helper here is mutable. Does it matter if the `SimulationTraceCheck` is checked in a different order?

What I mean is  the check `A-> B -> C`

A and C doesn't mutate helper while B mutate the helper. If I reorder the check like `B -> A -> C`. Does that matter to A while some state is changed in B?

Reply:
Same question to other checks.

Reply:
The order doesn't matter - can be in any way.

It is mutable because for different reason: for example it contains `code_hashes`, and multiple checks can do something with `code_hashes`. So if one check fetches `code_hashes` from eth_client, the other checks doesn't have to repeat the same calls (saving some time).

Reply:
Ok. Got you.